File: JDK11/opentraceability/src/main/java/opentraceability/models/events/kdes/EventKDEString.java
package opentraceability.models.events.kdes;

import opentraceability.interfaces.IEventKDE;
import opentraceability.utility.XAttribute;
import opentraceability.utility.XElement;
import org.json.JSONObject;
import org.w3c.dom.Attr;
import org.w3c.dom.Document;
import org.w3c.dom.Element;

import javax.xml.parsers.DocumentBuilderFactory;

import java.util.HashMap;
import java.util.Map;

public class EventKDEString extends IEventKDE {
    public Class valueType = String.class;
    public String value;
    public String type;
    public Map<String, String> attributes = new HashMap<>();

    public EventKDEString() {}

    public EventKDEString(String ns, String name) {
        this.namespace = ns;
        this.name = name;
        attributes = new HashMap<>();
    }

    public Object getJson() {
        if (value == null || value.isEmpty()) {
            return null;
        } else {
            JSONObject json = new JSONObject();
            json.put("keyName", value);
            return json;
        }
    }

    public XElement getXml() throws Exception {
        if (value == null || value.isEmpty()) {
            return null;
        } else {
            XElement element = new XElement(namespace, name);
            element.setValue(value);

            for (Map.Entry<String, String> entry : attributes.entrySet())
            {
                element.SetAttributeValue(null, entry.getKey(), entry.getValue());
            }

            return element;
        }
    }

    public void setFromJson(Object json) {
        attributes = new HashMap<>();
        value = json.toString();
    }

    public void setFromXml(XElement xml) throws Exception {
        value = xml.getValue();
        namespace = xml.getNamespaceUri();
        name = xml.getTagName();

        attributes = new HashMap<>();
        for (XAttribute xatt: xml.Attributes()) {
            attributes.put(xatt.Name, xatt.Value);
        }
    }

    @Override
    public String toString() {
        return value != null ? value : "";
    }

}

File: Java/OpenTraceability/src/main/opentraceability/models/events/kdes/eventkdestring.kt
package models.events.kdes

import interfaces.IEventKDE
import org.json.simple.JSONObject
import java.lang.reflect.Type
import org.w3c.dom.Element
import org.w3c.dom.Document
import javax.xml.parsers.DocumentBuilderFactory

class EventKDEString: EventKDEBase, IEventKDE {

    override var ValueType: Type = String::class.java

    var Value: String? = null


    var Type: String? = null
    var Attributes: MutableMap<String, String> = mutableMapOf()


    constructor(ns: String, name: String) {
        this.Namespace = ns;
        this.Name = name;
    }


    override fun GetJson(): JSONObject? {
        if (this.Value.isNullOrBlank()) {
            return null
        } else {
            val json = JSONObject()
            json.put("keyName", this.Value)
            return json
        }
    }

    override fun GetXml(): Element? {
        if (this.Value.isNullOrBlank()) {
            return null
        } else {
            val docFactory = DocumentBuilderFactory.newInstance()
            val docBuilder = docFactory.newDocumentBuilder()
            val document: Document = docBuilder.newDocument()

            val element = document.createElementNS(this.Namespace, this.Name)
            element.textContent = this.Value

            // set the xsi type...
            for ((key, value) in this.Attributes) {
                val attr = document.createAttributeNS(this.Namespace, key)
                attr.value = value
                element.setAttributeNodeNS(attr)
            }

            return element
        }
    }

    fun setFromJson(json: JToken) {
        this.Value = json.toString()
    }

    fun setFromXml(xml: XElement) {
        this.Value = xml.value

        for (xatt in xml.attributes()) {
            Attributes[xatt.name.toString()] = xatt.value
        }
    }

    override fun toString(): String {
        return Value ?: ""
    }

}
